Make Tooltip upgradable

reaperrr opened the pull request:
This allows the tooltip to change via upgrades, for example when an actor is deployed or reaches elite status.

reaperrr committed:
Made Tooltip trait upgradable

reaperrr committed:
Update tooltip when Tick Tank is deployed

pchote commented on OpenRA.Game/Traits/Player/FrozenActorLayer.cs:
```diff
@@ -80,7 +80,7 @@ public FrozenActor(Actor self, PPos[] footprint, Shroud shroud, bool startsRevea
 			TargetTypes = self.GetEnabledTargetTypes().ToHashSet();
 
-			tooltip = self.TraitsImplementing<ITooltip>().FirstOrDefault();
+			tooltip = self.TraitsImplementing<ITooltip>().FirstOrDefault(Exts.IsTraitEnabled);
```
You'll need to store all of these here, and then pick the appropriate one in `RefreshState`.

pchote commented on mods/ts/rules/nod-vehicles.yaml:
```diff
@@ -73,6 +73,13 @@ TTNK:
+		UpgradeTypes: deployed
+		UpgradeMaxEnabledLevel: 0
+	Tooltip@DEPLOYED:
+		Name: Deployed Tick Tank
```
"Tick Tank (deployed)" please

MustaphaTR replied:
Deployed Tick Tank is what vanilla TS uses. But i think Tick Tank (deployed)  is better too.

reaperrr commented:
Updated.

pchote commented on mods/ts/rules/nod-vehicles.yaml:
```diff
@@ -73,6 +73,13 @@ TTNK:
+		UpgradeMaxEnabledLevel: 0
+	Tooltip@DEPLOYED:
+		Name: Tick Tank (deployed)
+		Description: Nod's main battle tank.\nDeployed to gain extra protection.\n  Strong vs Vehicles\n  Weak vs Infantry, Aircraft
```
This highlights the bogosity of having this on `Tooltip` instead of `Buildable`.  These alternate descriptions will never be used.  I won't ask you to do it here (it belongs in its own PR), but we really need to move this to the proper trait.

reaperrr replied:
Filed #12045 so we don't forget.

reaperrr commented:
I think it's better to get #12141 merged first.

reaperrr commented:
Rebased.

reaperrr commented on OpenRA.Mods.Common/Widgets/ObserverProductionIconsWidget.cs:
```diff
@@ -91,7 +91,7 @@ public override void Draw()
 				var rsi = actor.TraitInfo<RenderSpritesInfo>();
 				var icon = new Animation(world, rsi.GetImage(actor, world.Map.Rules.Sequences, faction));
-				icon.Play(actor.TraitInfo<TooltipInfo>().Icon);
+				icon.Play(actor.TraitInfos<TooltipInfo>().First(Exts.IsTraitEnabled).Icon);
```
I was just wondering, shouldn't we move `Icon` to `Buildable` as well?

reaperrr commented:
Filed #12258 which should be merged first, too.

reaperrr commented on OpenRA.Mods.Common/Traits/World/EditorActorPreview.cs:
```diff
@@ -70,7 +70,9 @@ public EditorActorPreview(WorldRenderer worldRenderer, string id, ActorReference
 
-			var tooltip = Info.TraitInfoOrDefault<EditorOnlyTooltipInfo>() as TooltipInfoBase ?? Info.TraitInfoOrDefault<TooltipInfo>();
+			var tooltip = Info.TraitInfos<EditorOnlyTooltipInfo>().FirstOrDefault() as TooltipInfoBase
+				?? Info.TraitInfos<TooltipInfo>().FirstOrDefault();
```
Note to self: This should probably be `.First(Exts.IsTraitEnabled)` too, in any case it should match `ActorSelectorLogic`.

abcdefg30 commented:
The dependencies are merged now.

reaperrr commented:
Rebased and updated.

pchote commented on OpenRA.Game/Traits/Player/FrozenActorLayer.cs:
```diff
@@ -101,8 +101,9 @@ public void RefreshState()
 			}
 
-			if (tooltip != null)
+			if (tooltips.Any(Exts.IsTraitEnabled))
```
You can make this cleaner by pulling a `var tooltip = tooltips.FirstOrDefault(Exts.IsTraitEnabled)` outside the conditional and then keeping this as a null check.

pchote commented on OpenRA.Mods.Common/Traits/Tooltip.cs:
```diff
@@ -13,11 +13,11 @@
 		[Translate] public readonly string Name = "";
 
-		public abstract object Create(ActorInitializer init);
+		public override object Create(ActorInitializer init) { return this; }
```
Returning `this` here is bogus (that is a traitinfo, not a trait).  You should be able to remove this line completely.

pchote commented on OpenRA.Mods.Common/Widgets/Logic/Ingame/ProductionTooltipLogic.cs:
```diff
@@ -115,7 +115,7 @@ static string ActorName(Ruleset rules, string a)
 			ActorInfo ai;
 			if (rules.Actors.TryGetValue(a.ToLowerInvariant(), out ai) && ai.HasTraitInfo<TooltipInfo>())
-				return ai.TraitInfo<TooltipInfo>().Name;
+				return ai.TraitInfos<TooltipInfo>().First(Exts.IsTraitEnabled).Name;
```
It looks like this method is expected to return `a` if there isn't a valid tooltip, so better to use `FirstOrDefault` and only return if non-null here.

pchote commented on OpenRA.Mods.Common/Widgets/ViewportControllerWidget.cs:
```diff
@@ -191,7 +191,7 @@ public void UpdateMouseover()
 			if (underCursor != null)
 			{
-				ActorTooltip = underCursor.TraitsImplementing<ITooltip>().First();
+				ActorTooltip = underCursor.TraitsImplementing<ITooltip>().First(Exts.IsTraitEnabled);
```
Crashing here if all tooltips are disabled isn't great (previously the check in `underCursor` would have prevented this.  It would be better to use `FirstOrDefault` and then only set `ActorTooltipExtra` / `TooltipType` if it is non-null.

reaperrr committed:
Show internal name instead of crashing when buildable actor has no tooltip

But throw a lint error instead.

reaperrr commented:
Updated accordingly.

In addition, I changed `Buildable` actors without `Tooltip` to not crash the game, but display the internal name on the build palette instead and throw a lint error.
I believe this is a bit more modder-friendly for mods in their early stages, as actors won't require `Tooltip` to be buildable.

GraionDilach commented:
Please add an upgraded tooltip to the Sensor Array as well.

reaperrr committed:
Upgrade tooltip when Mobile Sensor Array is deployed

reaperrr commented:
> Please add an upgraded tooltip to the Sensor Array as well.

Updated.

abcdefg30 commented on OpenRA.Mods.Common/Widgets/Logic/Ingame/ProductionTooltipLogic.cs:
```diff
@@ -115,7 +115,11 @@ static string ActorName(Ruleset rules, string a)
 		{
 			ActorInfo ai;
 			if (rules.Actors.TryGetValue(a.ToLowerInvariant(), out ai) && ai.HasTraitInfo<TooltipInfo>())
```
Can't we drop the `HasTraitInfo` check now?

reaperrr replied:
Done.

abcdefg30 commented:
Looks good to me otherwise. :+1: / :white_check_mark:

reaperrr committed:
Remove now-redundant TooltipInfo check

reaperrr commented:
Updated.

obrakmann reviewed (approved):
👍

obrakmann commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/12729333016d5b6ef42830b9a8e135ce262fe830)
